docs/architecture/adr-013-metrics.md

```diff
@@ -10,85 +10,145 @@ Proposed
+    return m.gatherPrometheus()
+
+  case FormatText:
+    return m.gatherGeneric()
+
+  case FormatDefault:
+    return m.gatherGeneric()
+
+  default:
+    return GatherResponse{}, fmt.Errorf("unsupported metrics format: %s", format)
+  }
 }
 ```
 
-Each module can define its own `Metrics` type and`CreateMetrics` function in the x/\<module\>/observability/metrics.go file:
+In addition, `Metrics` allows us to gather the current set of metrics at any given point in time. An
+operator may also choose to send a signal, SIGUSR1, to dump and print formatted metrics to STDERR.
+
+During an application's bootstrapping and construction phase, if `Telemetry.Enabled` is `true`, the
+API server will create an instance of a reference to `Metrics` object and will register a metrics
+handler accordingly.
 
 ```go
-type Metrics struct {
-  Size metrics.Guage
+func (s *Server) Start(cfg config.Config) error {
+  // ...
 
-  Transactions metrics.Counter
-}
+  if cfg.Telemetry.Enabled {
+    m, err := telemetry.New(cfg.Telemetry)
+    if err != nil {
+      return err
+    }
 
-func CreateMetrics(namespace string, labelsAndValues... string) *Metrics {
-  labels := make([]string, len(labelsAndValues/2))
-  for i := 0; i < len(labelsAndValues); i += 2 {
-      labels[i/2] = labelsAndValues[i]
+    s.metrics = m
+    s.registerMetrics()
   }
-  return &Metrics{
-    Size: prometheus.NewGaugeFrom(stdprometheus.GaugeOpts{
-		Namespace: namespace,
-		Subsystem: "subsystem",
-		Name:      "size",
-		Help:      "Size of the custom metric",
-	}, labels).With(labelsAndValues...),
-    Transactions: prometheus.NewCounterFrom(stdprometheus.CounterOpts{
-                Namespace: namespace,
-                Subsystem: "subsystem",
-                Name:      "transactions",
-                Help:      "Number of transactions processed",
-        }, labels).With(labelsAndValues...),
+
+  // ...
+}
+
+func (s *Server) registerMetrics() {
+  metricsHandler := func(w http.ResponseWriter, r *http.Request) {
+    format := strings.TrimSpace(r.FormValue("format"))
+
+    gr, err := s.metrics.Gather(format)
+    if err != nil {
+      rest.WriteErrorResponse(w, http.StatusBadRequest, fmt.Sprintf("failed to gather metrics: %s", err))
+      return
+    }
+
+    w.Header().Set("Content-Type", gr.ContentType)
+    _, _ = w.Write(gr.Metrics)
   }
 
+  s.Router.HandleFunc("/metrics", metricsHandler).Methods("GET")
+}
 ```
 
-To get the correct namespace for the modules changing `BasicManager` to consist of the app name is needed.
+Application developers may track counters, gauges, summaries, and key/value metrics. There is no
+additional lifting required by modules to leverage profiling metrics. To do so, it's as simple as:
 
 ```go
-type BasicManager struct {
-	appName string
-	modules map[string]AppModuleBasic
+func (k BaseKeeper) MintCoins(ctx sdk.Context, moduleName string, amt sdk.Coins) error {
+  defer metrics.MeasureSince([]string{"MintCoins"}, time.Now().UTC())
```

Comment:
I'm a little confused as to how this works. Is `metrics` the alias for the go-metrics package here? And does the library work using globally scoped functions?

Reply:
https://github.com/cosmos/cosmos-sdk/pull/6399/files#diff-78855aa869d1781cea93a2f0dbbe35d1R90-R92

Reply:
I will probably create auxiliary methods in this package to make sure proper labels are added. e.g. `telemetry.ModuleMeasureSince(moduleName, ...)`.